### PR TITLE
core/exec: do not crash with UtmpMode=user without User= setting

### PR DESCRIPTION
**Pull request by yuwata**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=2264404.

Replaces #31356.

### CODE REVIEW

File: src/core/exec-invoke.c [hunk 1]
```diff
@@ -4451,6 +4451,17 @@ int exec_invoke(
...
         if (context->utmp_id) {
+                _cleanup_free_ char *username_alloc = NULL;
+
+                if (!username && context->utmp_mode == EXEC_UTMP_USER) {
+                        username_alloc = uid_to_name(uid_is_valid(uid) ? uid : saved_uid);
+                        if (!username_alloc) {
+                                *exit_status = EXIT_USER;
+                                return log_oom();
+                        }
+                        username = username_alloc;
```
Comment (YHNdnzj) [inline src/core/exec-invoke.c]: No. `username` corresponds to the `User=` setting. This means that we assume User= is set in `build_environment` and possibly other places, breaking the semantics of e.g. `SetLoginEnvironment=`

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: src/core/exec-invoke.c (hunk 1 above)
Comment (YHNdnzj) [reply src/core/exec-invoke.c]: And `username` becomes dangling when username_alloc gets destroyed?

Comment (yuwata) [reply src/core/exec-invoke.c]: > And `username` becomes dangling when username_alloc gets destroyed?

Indeed. Fixed.

**Review by keszybz — APPROVED** [review]
LGTM.

**Comment by poettering** [comment]
lgtm

**Pull request merged**